packages/vite/src/node/plugins/html.ts

```diff
@@ -955,11 +955,21 @@ export function htmlEnvHook(config: ResolvedConfig): IndexHtmlTransformHook {
   const pattern = /%(\S+?)%/g
   const envPrefix = resolveEnvPrefix({ envPrefix: config.envPrefix })
   const env: Record<string, any> = { ...config.env }
+
   // account for user env defines
   for (const key in config.define) {
     if (key.startsWith(`import.meta.env.`)) {
       const val = config.define[key]
-      env[key.slice(16)] = typeof val === 'string' ? val : JSON.stringify(val)
+      if (typeof val === 'string') {
+        try {
+          const parsed = JSON.parse(val)
+          env[key.slice(16)] = typeof parsed === 'string' ? parsed : val
```

Comment:
I totally forgot about this PR 😅
I replaced `extractStringLiteral` with this one.
The `typeof parsed === 'string'` thing is to avoid a breaking change when define has something like `'import.meta.env.VITE_OBJECT_STRING': '{ "foo": "bar" }'`.